firedrake/assemble.py

```diff
@@ -522,21 +529,44 @@ def base_form_assembly_visitor(self, expr, tensor, *args):
             if all(isinstance(op, numbers.Complex) for op in args):
                 result = sum(weight * arg for weight, arg in zip(weights, args))
                 return tensor.assign(result) if tensor else result
-            elif (all(isinstance(op, firedrake.Cofunction) for op in args)
+
+            # Compress repeated terms
+            terms = defaultdict(int)
```

Comment:
I reckon it would be clearer to instead partition the sum into tensor bits and non-tensor bits

Reply:
I clarified the logic in the comments.